### PR TITLE
Add a command for building student notebooks and autograder scripts.

### PR DESCRIPTION
**Pull request by salikh**

### CODE REVIEW

File: go/notebook/README.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
+# Assignment builder
+
+This directory contains code for the assignment builder, which is a tool that
+takes a master notebook as input, and produces a student notebook as well as a
+autograder directory with the tests.
+
+This tool is very similar to jassign (https://github.com/okpy/jassign), but the
+requirement set is slightly different, and jassign seems to be a quite young
+project, so it did not look appropriate to reuse . However, given the goals are
```
Comment (victorkwan) [inline go/notebook/README.md]:
```suggestion
project, so it did not look appropriate to reuse. However, given the goals are
```

File: go/notebook/README.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      # BEGIN SOLUTION
+      return 2
+      # END SOLUTION
+
+TODO(salikh): Implement a special marker (`# UNITTEST OUTPUT`?) that instructs
+that the stdout output of the master solution should be recorded and used for
+testing student submissions.
+
+The solutions are removed when producing a student notebook, with the
+replacement being either a generic "...", or provided with `PROMPT` markers.
```
Comment (victorkwan) [inline go/notebook/README.md]: Would it make more sense just to have a dummy comment like the one below in the absence of `PROMPT` markers?

File: go/notebook/README.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+Autograder tests are performed by providing an potentially incorrect submission
+using `%%submission` cell magic, and running the unit test using `autotest`
+funciont. The resulting outcome vector can be checked using plain Python
+`assert` statements.
+
+### Report scripts
+
+Report scripts are used by the autograder to provide human-readable feedback
+without necesserily revealing the autograder tests themselves.
```
Comment (victorkwan) [inline go/notebook/README.md]:
```suggestion
without necessarily revealing the autograder tests themselves.
```

File: go/notebook/README.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+Note: to test the autograder scripts, the second-meta-level tests provide
+various incorrect inputs, and the unit tests are run and the outcome vector is
+checked against the expected one.
+
+### Autograder tests (self-tests)
+
+Autograder tests are performed by providing an potentially incorrect submission
+using `%%submission` cell magic, and running the unit test using `autotest`
+funciont. The resulting outcome vector can be checked using plain Python
```
Comment (victorkwan) [inline go/notebook/README.md]:
```suggestion
functionality. The resulting outcome vector can be checked using plain Python
```

File: go/notebook/README.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+These tests are intended to be used for autograding, i.e. running the tests on
+potentially incorrect solution in order to identify problem points and give some
+feedback to students. They are marked with `UNITTEST` and are used in two ways:
+
+*   To test the master solution (all unit tests should pass on the master
+    solution).
+*   To extract the autograder scripts.
+
+Note: to test the autograder scripts, the second-meta-level tests provide
```
Comment (victorkwan) [inline go/notebook/README.md]: This and the following section aren't as obvious as the previous ones. Can we provide an example for what an autograder test/script might look like?

File: go/notebook/README.md (hunk 2 above)
Comment (salikh) [reply go/notebook/README.md]: I thought of having a more detailed dummy comment, but realized that it's hard to pick a good one (e.g. something that would be good both for Japanese and English), so eventually decided on using just "..." by default.

File: go/notebook/README.md (hunk 5 above)
Comment (salikh) [reply go/notebook/README.md]: Added a pointer to standalone examples.
I will need to revise the documentation and examples after integrating the recent %%inlinetest changes, and after that we need to consider how to achieve both performance and flexibility for the tests. I'll talk to you later.

**Comment by salikh** [comment]
Merged directly.

**Pull request closed without merging**